src/builtins/alias.c

```diff
@@ -0,0 +1,104 @@
+/*
+    psh/backends/posix2/run.c - process-related posix backend
+    Copyright 2020 Manuel Bertele
+
+    This file is part of Psh, P shell.
+
+    Psh is free software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    Psh is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with this program.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include "backend.h"
+#include "builtin.h"
+#include "libpsh/util.h"
+#include "libpsh/xmalloc.h"
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+static int add_alias(psh_hash *table, char *alias, char *value)
+{
+
+    psh_hash_add_chk(table, alias, value, 1);
+
+    return 0;
+}
+
+static char *check_for_alias(psh_hash *table, char *alias)
+{
+    return psh_hash_get(table, alias);
+
+}
+
+
+
+int builtin_alias(int argc, char **argv, psh_state *state)
+{
+
+    int i;
+    char *alias, *value, *argstr;
+
+    argstr = strdup(argv[1]);
+
+    /* Combine all argv pointers to one pointer */
+    for (i = 2; i < argc; i++)
+    {
+        xrealloc(argstr, strlen(argstr) + strlen(argv[i]) + 1);
```

Comment:
If you want to use `realloc` (or `xrealloc`, doesn't matter), remember to assign the return value back to the pointer, e.g.
```
argstr = xrealloc(argstr, newlen);
```
Because the function CANNOT change what your pointer points to without you assigning, but it might change the location of the allocated block, since the original one might not be big enough. A little code to test this is:
```
char *newp, *p = malloc(1); /* Small allocation */
printf("malloc(1) returned: %p\n", p);
newp = realloc(p, 1000000); /* Bigger allocation, approx. 1M */
printf("realloc moved the data to a new location of %p, but the old pointer is still at %p\n", newp, p);
```
On my computer, it yielded:
```
malloc(1) returned: 0x7fffd1f30260
realloc moved the data to a new location of 0x7ff3ed360010, but the old pointer is still at 0x7fffd1f30260
```